internal/state/store.go

```diff
@@ -408,60 +420,65 @@ func (store dbStore) reverseBatchDelete(batch dbm.Batch, start, end []byte) ([]b
 
 //------------------------------------------------------------------------
 
-// LoadABCIResponses loads the ABCIResponses for the given height from the
-// database. If not found, ErrNoABCIResponsesForHeight is returned.
+// LoadFinalizeResponses loads the responses to FinalizeBlock for the given height from the
+// database. If not found, ErrNoFinalizeResponsesForHeight is returned.
 //
 // This is useful for recovering from crashes where we called app.Commit and
 // before we called s.Save(). It can also be used to produce Merkle proofs of
 // the result of txs.
-func (store dbStore) LoadABCIResponses(height int64) (*tmstate.ABCIResponses, error) {
-	buf, err := store.db.Get(abciResponsesKey(height))
+func (store dbStore) LoadFinalizeResponses(height int64) (*abci.ResponseFinalizeBlock, error) {
+	buf, err := store.db.Get(finalizeResponsesKey(height))
 	if err != nil {
 		return nil, err
 	}
 	if len(buf) == 0 {
-
-		return nil, ErrNoABCIResponsesForHeight{height}
+		return nil, ErrNoFinalizeResponsesForHeight{height}
+	}
+	if len(buf) == 1 && buf[0] == emptyResponsesMagic {
```

Comment:
Please let me know if there is a more elegant way of doing this.
AFAIU, we MUST distinguish whether results were stored for a height (even if empty), or not. If we remove this distinction, we lose important information when it comes to troubleshooting (bad?) usage of the store.

An alternative could be to error out is trying to save empty responses to the store. However, this will break UTs that are explicitly requiring this work... and there might be similar existing Apps out there. The safest is always to keep _exact_ functionality in infra (such as the store)... especially if there is a UT explicitly demanding it.

Reply:
The only usage of this call is for reindexing events and recovering from a crash right? I'm not sure if there is actually any value in keeping the distinction between nil and empty

Reply:
The thing that made me introduce the 1-byte-hack to keep the distinction is `TestFinalizeResponsesSaveLoad2`.
This TC is insistingly testing various kinds of `nil` in the `ExecTxResults`, and, without the hack, the test is failing.

git blame is confusing because of a revert, but after some git kung-fu I got to commits 89cdde7 and 73fb1c3a17, which don't tell much either. Anyway these are more than 4 years old so I doubt the authors remember _why_ those nils.

CONCLUSION: If we are sure the distinction is unnecessary then I'll refactor the test case and remove the hack.

Reply:
Agreed during team meeting to remove the hack, and adapt the tests that don't like `SaveFinalizeBlockResponses` returning an error when the serialized buffer has 0 bytes